internal/controllers/watcher/controller.go

```diff
@@ -0,0 +1,170 @@
+package watcher
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"open-cluster-management.io/addon-framework/pkg/addonmanager"
+	addonapiv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	workapiv1 "open-cluster-management.io/api/work/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var noReconcilePred = builder.WithPredicates(predicate.Funcs{
+	UpdateFunc:  func(ue event.UpdateEvent) bool { return false },
+	CreateFunc:  func(e event.CreateEvent) bool { return false },
+	DeleteFunc:  func(e event.DeleteEvent) bool { return false },
+	GenericFunc: func(e event.GenericEvent) bool { return false },
+})
+
+type WatcherManager struct {
+	mgr    *ctrl.Manager
+	logger logr.Logger
+}
+
+func NewWatcherManager(logger logr.Logger, scheme *runtime.Scheme, addonManager *addonmanager.AddonManager) (*WatcherManager, error) {
+	l := logger.WithName("mcoa-watcher")
+
+	ctrl.SetLogger(l)
+
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{Scheme: scheme})
+	if err != nil {
+		return nil, fmt.Errorf("unable to start manager: %w", err)
+	}
+
+	if err = (&WatcherReconciler{
+		Client:        mgr.GetClient(),
+		Log:           l.WithName("controllers").WithName("mcoa-watcher"),
+		Scheme:        mgr.GetScheme(),
+		addonnManager: addonManager,
+	}).SetupWithManager(mgr); err != nil {
+		return nil, fmt.Errorf("unable to create mcoa-watcher controller: %w", err)
+	}
+
+	if err = mgr.AddHealthzCheck("health", healthz.Ping); err != nil {
+		return nil, fmt.Errorf("unable to set up health check: %w", err)
+	}
+	if err = mgr.AddReadyzCheck("check", healthz.Ping); err != nil {
+		return nil, fmt.Errorf("unable to set up ready check: %w", err)
+	}
+
+	wm := WatcherManager{
+		mgr:    &mgr,
+		logger: l,
+	}
+
+	return &wm, nil
+}
+
+func (wm *WatcherManager) Start(ctx context.Context) {
+	wm.logger.Info("Starting watcher manager")
+	go func() {
+		err := (*wm.mgr).Start(ctx)
+		if err != nil {
+			wm.logger.Error(err, "there was an error while running the reconciliation watcher")
+		}
+	}()
```

Comment:
Do we need to use some sort of context cancelation here to avoid leaking go routines? :thinking:

Reply:
I'm not sure about that because... the goroutine will be blocked here:
```go
		err := (*wm.mgr).Start(ctx)
```
If something stops the manager from outside, the routine will finish too, right?

Reply:
Yeah what's the outside here? Does the start method listen for interrupt signals internally? I guess so because that's the same method used in regular operators. However it would be good to double check because this method is not usually enclosed in a go routine. It's typically part of main and spawns go routines itself

Reply:
> Yeah what's the outside here? Does the start method listen for interrupt signals internally? 

It does.